timemachine/potentials/jax_utils.py

```diff
@@ -36,19 +89,74 @@ def rescale_coordinates(
 
     return conf + offset[indices]
 
-
 def delta_r(ri, rj, box=None):
     diff = ri - rj # this can be either N,N,3 or B,3
     dims = ri.shape[-1]
 
     # box is None for harmonic bonds, not None for nonbonded terms
     if box is not None:
-        for d in range(dims):
-            diff -= box[d]*np.floor(np.expand_dims(diff[...,d], axis=-1)/box[d][d]+0.5)
+
+        # Equivalent to:
+        #    for d in range(dims):
+        #       diff -= box[d]*np.floor(np.expand_dims(diff[...,d], axis=-1)/box[d][d]+0.5)
+        def loop_body(d, diff):
+            return diff - box[d] * np.floor(np.expand_dims(diff[..., d], axis=-1) / box[d][d] + 0.5)
+        diff = jax.lax.fori_loop(lower=0, upper=dims, body_fun=loop_body, init_val=diff)
 
     return diff
 
 
+def distance_on_pairs(ri, rj, box=None):
+    """O(n) where n = len(ri) = len(rj)"""
+    assert len(ri) == len(rj)
+
+    diff = delta_r(ri, rj, box)
+    dij = np.linalg.norm(diff, axis=-1)
+
+    assert len(dij) == len(ri)
+
+    return dij
+
+
+def neighbor_mask(conf, inds_l, inds_r, cutoff, box):
+    dij = distance_on_pairs(conf[inds_l], conf[inds_r], box)
+    return dij < cutoff
+
+
+def batched_neighbor_inds(confs, inds_l, inds_r, cutoff, boxes):
+    """Given candidate interacting pairs (inds_l, inds_r),
+        inds_l.shape == n_interactions
+    exclude most pairs whose distances are >= cutoff (neighbor_inds_l, neighbor_inds_r)
+        neighbor_inds_l.shape == (len(confs), max_n_neighbors)
+        where the total number of neighbors returned for each conf in confs is the same
+        max_n_neighbors
+
+    This padding causes some amount of wasted effort, but keeps things nice and fixed-dimensional
```

Comment:
Can you confirm that (None,) leading dimensions still trigger a separate JIT pass? I thought if the first dim were variadic, the CUDA back-end at least would avoid repeated jitting.

Reply:
Sorry, I'm not sure how to confirm this! Is there an example you can point me to, illustrating where to provide (None,) leading dimensions?

[This comment](https://github.com/google/jax/issues/2521#issuecomment-604759386) on the Jax issue tracker suggested that it does need to be shape-specialized for now, and later comments suggest that the Jax masking approach to emulating variable-length behavior with fixed-length arrays is being reworked.

Reply:
You're right - XLA doesn't support None still. Oof.